### PR TITLE
fix contour boundingbox with `NaN`s

### PR DESCRIPTION
**Pull request by t-bltg**
# Description

Fixes [discord issue](https://discord.com/channels/996787732149981214/1148180700005154876/1148180700005154876) (fix https://github.com/MakieOrg/Makie.jl/issues/3211):

```julia
let
    a = b = 1:3
    c = [0 1 2; 1 2 3; 4 5 6]
    ct1 = contour(a, b, c; levels = [1, 2, 3], labels = true)   # OK
    d = [0 1 2; 1 2 3; 4 5 NaN]
    ct2 = contour(a, b, d; levels = [1, 2, 3], labels = true)   # NOK
end
```

Fix a `CairoMakie` bug when glyphs positions contained a `NaN`, corrupting the Cairo drawing state / context.

## Type of change

Delete options that do not apply:

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [ ] Added an entry in NEWS.md (for new features and breaking changes)
- [ ] Added or changed relevant sections in the documentation
- [x] Added unit tests for new algorithms, conversion methods, etc.
- [ ] Added reference image tests for new plotting functions, recipes, visual options, etc.

### CODE REVIEW

**Comment by Lightup1** [comment]
Hi @t-bltg ,
I tried this branch.
There is no error message now, but there are also no contour lines now.

**Comment by Lightup1** [comment]
image

**Comment by jkrumbiegel** [comment]
Which backend was this?

**Comment by Lightup1** [comment]
@jkrumbiegel CairoMakie

**Comment by t-bltg** [comment]
Let me have a look this morning.

**Comment by jkrumbiegel** [comment]
Could be a bug in CairoMakie, too, because there is one little line segment and afterwards nothing, also the axis lines are missing. That usually happens when Cairo silently corrupts the surface state, which can happen with NaNs sometimes.

**Comment by t-bltg** [comment]
Let's reason first with `Nan`-less arrays: `c = [0 1 2; 1 2 3; 4 5 6]`.

What is visually represented is `reverse(c; dims=1)`:
```
3×3 Matrix{Float64}:
 4.0  5.0    6.0
 1.0  2.0    3.0
 0.0  1.0    2.0
```

So this is `contourf(a, b, c; levels)`:
contourf

With `GLMakie` (or `CairoMakie`), `contour(a, b, c; levels, labels=true)`:
glmakie.

Now with `NaN`s, `d = [0 1 2; 1 2 3; 4 5 NaN]`, `contourf(a, b, d; levels)`:
contourf-nan.

So I will expect at least the `1` and `2` levels lines with labels to appear. I'll rework the PR.

**Comment by Lightup1** [comment]
GLMakie works fine now. But CairoMakie produces no contour lines.

**Comment by t-bltg** [comment]
With GLMakie:

glmakie

Lines, but missing 2 contour labels.

**Comment by t-bltg** [comment]
I'm marking as ready for review because I have no ideas left currently.
The contour algorithm output is different whether there is a `NaN` or not.

**Comment by Lightup1** [comment]
Dear @t-bltg ,
Is there any quick fixing way to make Makie ignore NaN?
I tried to replace NaN with 1e6 in some cases it works fine, but in some other cases (when the border is smaller than the specified levels) it will draw multiple lines near the border between normal values and NaN.
Does julia support void element in a matrix which is ignored by Makie?

**Comment by t-bltg** [comment]
> Does julia support void element in a matrix which is ignored by Makie?

I would want `missing` (or `nothing`) to behave how you describe unfortunately it doesn't work either in this case.

**Comment by jkrumbiegel** [comment]
Aside from the NaN-interpolation question discussed above, is this now in mergeable state? With NaN causing the appropriate gaps?

**Comment by t-bltg** [comment]
Current state, output from the code snippet in https://github.com/MakieOrg/Makie.jl/pull/3210#issue-1880608285.

**CairoMakie - no Nans**
cairo_no_nans

**CairoMakie - Nans**
cairo_nans

**GLMakie - no Nans**
gl_no_nans

**GLMakie - Nans**
gl_nans

**Comment by jkrumbiegel** [comment]
Ok so CairoMakie is still buggy

**Comment by t-bltg** [comment]
Is inconsistency between `CairoMakie` and `GLMakie` backends happening at backend level (the way every backend deals with segments) or is this something we can solve in `Makie` ?

**Comment by jkrumbiegel** [comment]
This must be a NaN-handling bug in CairoMakie. It may not actually _draw_ NaNs, they have to be skipped (but one has to be careful to create the right paths in the process). And here it seems it's drawn (or something else that doesn't work) because the axis lines are missing, so Cairo broke in the process of drawing this.

**Comment by t-bltg** [comment]
Oh right I didn't notice the axis lines missing.
Do you have any hint where this might occur ?
I'll investigate before the end of the week ...

**Comment by t-bltg** [comment]
@jkrumbiegel, this is the new output, the bug was when drawing a glyph with `NaN`s.

**GLMakie - no NaNs c**
gl-no-nans-c
**GLMakie - NaNs d**
gl-nans-d

**CairoMakie - no NaNs c**
cairo-no-nans-c

**CairoMakie - NaNs d**
cairo-nans-d

File: CairoMakie/src/primitives.jl [hunk 1]
```diff
@@ -587,6 +565,11 @@ function draw_glyph_collection(
         # offsets and scale apply in markerspace
         gp3 = glyph_pos[Vec(1, 2, 3)] ./ glyph_pos[4] .+ model33 * (glyphoffset .+ p3_offset)
 
+        if any(isnan, gp3)
```
Comment (t-bltg) [inline CairoMakie/src/primitives.jl:568]: This is the fix for https://github.com/MakieOrg/Makie.jl/pull/3210#issuecomment-1778933826.

**Comment by jkrumbiegel** [comment]
Could you remove all the non functionality related formatting changes? Makes it harder to see what's up. Also the empty ifs with comments can be left as they are, I know they technically don't do anything but they helped me understand what the code was doing.

**Comment by t-bltg** [comment]
> Could you remove all the non functionality related formatting changes?

Done.

> Makes it harder to see what's up.

I agree for the review, although these changes were make to simplify and make the current code consistent.

> Also the empty ifs with comments can be left as they are

Done.

**Review by jkrumbiegel — APPROVED** [review]

**Comment by jkrumbiegel** [comment]
Thank you! Looks good to me

**Pull request merged**